File: erpnext/projects/report/project_progress_report/project_progress_report.js
// Copyright (c) 2025, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Project Progress Report"] = {
	"filters": [
		{
			"fieldname": "parent_project",
			"label": ("Parent Project"),
			"fieldtype": "Select",
			"options": ["", "GI Project", "DFG Project", "Other Project"],
			"on_change": function(query_report) {
				frappe.query_report.set_filter_value("project_definition", null);
				frappe.query_report.set_filter_value("project", null);
				// frappe.query_report.set_filter_value("task", null);
				frappe.query_report.refresh();
			},
			"reqd": 1
		},
		{
			"fieldname": "project_definition",
			"label": ("Gyalsung Academy(Project Definition)"),
			"fieldtype": "Link",
			"options": "Project Definition",
			"get_query": function() {
				var parent_project = frappe.query_report.get_filter_value("parent_project")
				return {
					'doctype': "Project Definition",
					'filters': [['project_category', '=', parent_project]]
				}
			},
			"on_change": function(query_report) {
				frappe.query_report.set_filter_value("project", null);
				// frappe.query_report.set_filter_value("task", null);
				frappe.query_report.refresh();
			}
		},
		{
			"fieldname": "project",
			"label": ("Activity(Project)"),
			"fieldtype": "Link",
			"options": "Project",
			"get_query": function() {
				var project_definition = frappe.query_report.get_filter_value("project_definition")
				return { 'doctype': "Project",
						'filters': [
								['project_definition', '=', project_definition]
				]}
			},
			"on_change": function(query_report) {
				frappe.query_report.set_filter_value("task", null);
				frappe.query_report.refresh();
			}
		},
		// {
		// 	"fieldname": "task",
		// 	"label": ("Task"),
		// 	"fieldtype": "Link",
		// 	"options": "Task",
		// 	"get_query": function() {
		// 		var project = frappe.query_report.get_filter_value("project")
		// 		return { 'doctype': "Task",
		// 				'filters': [
		// 						['project', '=', project]
		// 		]}
		// 	},
		// },
	]
};

File: erpnext/projects/report/project_progress_report/project_progress_report.py
# Copyright (c) 2025, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import date_diff

def execute(filters=None):
	columns, data = [], []
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data

def get_columns(filters=None):
	columns = []
	if filters.get("parent_project") and not filters.get("project_definition") and not filters.get("project") and not filters.get("task"):
		columns = [
			{
			"fieldname": "project_definition",
			"label": "Project Definition",
			"fieldtype": "Link",
			"options": "Project Definition",
			"width": 150
			},
			{
			"fieldname": "start_date",
			"label": "Start Date",
			"fieldtype": "Date",
			"width": 120
			},
			{
			"fieldname": "end_date",
			"label": "End Date",
			"fieldtype": "Date",
			"width": 120
			},
			{
			"fieldname": "duration",
			"label": "Duration(Days)",
			"fieldtype": "Float",
			"width": 120
			},
			{
			"fieldname": "weightage",
			"label": "PP Weightage(%)",
			"fieldtype": "Float",
			"width": 120
			},
			{
			"fieldname": "site_progress",
			"label": "Site Progress(%)",
			"fieldtype": "Data",
			"width": 120
			},
			{
			"fieldname": "project_progress",
			"label": "Project Progress(%)",
			"fieldtype": "Float",
			"width": 120
			},
			{
			"fieldname": "estimated_budget",
			"label": "Provisional Estimated Budget",
			"fieldtype": "Float",
			"width": 120
			},
			{
			"fieldname": "actual_expenses",
			"label": "Actual Expense",
			"fieldtype": "Float",
			"width": 120
			},
			{
			"fieldname": "financial_progress",
			"label": "Financial Progress(%)",
			"fieldtype": "Float",
			"width": 120
			},
		]
	elif filters.get("parent_project") and filters.get("project_definition") and not filters.get("project") and not filters.get("task"):
		columns = [
			{
			"fieldname": "project",
			"label": "Project",
			"fieldtype": "Link",
			"options": "Project",
			"width": 150
			},
			{
			"fieldname": "start_date",
			"label": "Start Date",
			"fieldtype": "Date",
			"width": 120
			},
			{
			"fieldname": "end_date",
			"label": "End Date",
			"fieldtype": "Date",
			"width": 120
			},
			{
			"fieldname": "duration",
			"label": "Duration(Days)",
			"fieldtype": "Float",
			"width": 120
			},
			{
			"fieldname": "weightage",
			"label": "PP Weightage(%)",
			"fieldtype": "Float",
			"width": 120
			},
			{
			"fieldname": "site_progress",
			"label": "Site Progress(%)",
			"fieldtype": "Data",
			"width": 120
			},
			{
			"fieldname": "project_progress",
			"label": "Project Progress(%)",
			"fieldtype": "Float",
			"width": 120
			},
			{
			"fieldname": "estimated_budget",
			"label": "Provisional Estimated Budget",
			"fieldtype": "Float",
			"width": 120
			},
			{
			"fieldname": "actual_expenses",
			"label": "Actual Expenses",
			"fieldtype": "Float",
			"width": 120
			},
			{
			"fieldname": "financial_progress",
			"label": "Financial Progress(%)",
			"fieldtype": "Data",
			"width": 120
			},
		]
	return columns

def get_data(filters=None):
	data = []
	if filters.get("parent_project") and not filters.get("project_definition") and not filters.get("project") and not filters.get("task"):
		for pd in frappe.db.get_all("Project Definition", {"project_category": filters.get("parent_project"), "docstatus": 1}, ["name", "physical_progress", "physical_progress_weightage", "percent_completed" ,"provisional_estimated_budget", "a_expenses", "f_progress"]):
			start_date = None
			end_date = None
			duration = None
			start_dates = []
			end_dates = []
			cc_mandays = frappe.db.sql("""
                              select sum(mandays) from `tabProject` where project_definition = '{}' and docstatus < 2
                              """.format(pd.name))
			overall_mandays = frappe.db.sql("""
                              select sum(mandays) from `tabProject` where docstatus < 2
                              """.format(pd.name))
			for prj in frappe.db.get_all("Project Definition", {"name": pd.name}, ["start_date", "end_date"]):
				if prj.start_date:
					start_dates.append(prj.start_date)
				if prj.end_date:
					end_dates.append(prj.end_date)
			if len(start_dates) > 0:
				start_date = min(start_dates)
			if len(end_dates) > 0:
				end_date = min(end_dates)
			if start_date and end_date:
				duration = date_diff(end_date, start_date)+1
			data.append({"project_definition": pd.name, "start_date": start_date, "end_date": end_date, "duration": duration, "project_progress": pd.physical_progress, "weightage": pd.physical_progress_weightage, "site_progress": pd.percent_completed, "estimated_budget": pd.provisional_estimated_budget, "actual_expenses": pd.a_expenses, "financial_progress": pd.f_progress})
	elif filters.get("parent_project") and filters.get("project_definition") and not filters.get("project") and not filters.get("task"):
		for pd in frappe.db.get_all("Project", {"project_definition": filters.get("project_definition"), "docstatus": ["<", 2]}, ["name", "physical_progress", "physical_progress_weightage", "percent_completed", "total_duration", "expected_start_date", "expected_end_date", "estimated_budget", "actual_expenses", "financial_progress"]):
			start_date = None
			end_date = None
			duration = None
			# start_dates = []
			# end_dates = []
			# for prj in frappe.db.get_all("Project Definition", {"name": pd.name}, ["start_date", "end_date"]):
			# 	if prj.start_date:
			# 		start_dates.append(prj.start_date)
			# 	if prj.end_date:
			# 		end_dates.append(prj.end_date)
			# if len(start_dates) > 0:
			# 	start_date = min(start_dates)
			# if len(end_dates) > 0:
			# 	end_date = min(end_dates)
			# if start_date and end_date:
			# 	duration = date_diff(end_date, start_date)+1
			data.append({"project": pd.name, "start_date": pd.expected_start_date, "end_date": pd.expected_end_date, "duration": pd.total_duration, "project_progress": pd.physical_progress, "weightage": pd.physical_progress_weightage, "site_progress": pd.percent_completed, "estimated_budget": pd.estimated_budget, "actual_expenses": pd.actual_expenses, "financial_progress": pd.financial_progress})

	return data